File: golf-trends/lib/fetch-trends.ts
import { execFile } from 'child_process';
import path from 'path';
import type { RawTrendsRecord } from './types';

export async function runPythonFetch(): Promise<RawTrendsRecord> {
  const pythonPath = process.env.PYTHON_PATH || 'python3';
  const scriptPath = path.join(process.cwd(), 'scripts', 'fetch_trends.py');

  return new Promise((resolve, reject) => {
    execFile(
      pythonPath,
      [scriptPath],
      {
        timeout: 120_000, // 2 min — sequential fetches with delays
        maxBuffer: 1024 * 1024, // 1MB
        env: { ...process.env, PYTHONUNBUFFERED: '1' },
      },
      (error, stdout, stderr) => {
        if (error) {
          console.error('[fetch-trends] Python error:', error.message);
          if (stderr) console.error('[fetch-trends] stderr:', stderr.slice(0, 500));
          reject(new Error(`Python script failed: ${error.message}`));
          return;
        }
        if (stderr) {
          // pytrends logs to stderr — not fatal
          console.warn('[fetch-trends] Python stderr (non-fatal):', stderr.slice(0, 500));
        }
        try {
          const data = JSON.parse(stdout.trim()) as RawTrendsRecord;
          resolve(data);
        } catch (parseErr) {
          reject(new Error(`Failed to parse Python output: ${(parseErr as Error).message}`));
        }
      }
    );
  });
}

// Node.js fallback via google-trends-api npm package
// Used when Python/pytrends is unavailable.
export async function runNodeFetch(): Promise<RawTrendsRecord> {
  // Dynamic import so the server doesn't crash if package isn't installed
  const googleTrends = await import('google-trends-api').catch(() => null);
  if (!googleTrends) {
    throw new Error('google-trends-api not available — install it as a fallback');
  }

  const terms = [
    'golf clubs', 'golf balls', 'golf bags',
    'golf', 'golf equipment', 'golf simulator',
  ] as const;

  const results: RawTrendsRecord = {};
  const delay = (ms: number) => new Promise((r) => setTimeout(r, ms));

  for (const term of terms) {
    try {
      const raw = await googleTrends.default.interestOverTime({
        keyword: term,
        startTime: new Date('2017-01-01'),
        geo: 'US',
      });
      const parsed = JSON.parse(raw as string);
      const timeline: Array<{ time: string; value: number[] }> =
        parsed?.default?.timelineData ?? [];
      const record: Record<string, number> = {};
      for (const point of timeline) {
        const ts = new Date(parseInt(point.time, 10) * 1000).toISOString().split('T')[0];
        record[ts] = point.value[0] ?? 0;
      }
      results[term] = record;
      // Throttle
      await delay(2500 + Math.random() * 2000);
    } catch (err) {
      results[term] = { error: String(err) };
    }
  }

  return results;
}

File: golf-trends/scripts/fetch_trends.py
#!/usr/bin/env python3
"""
Golf Trends Fetcher — fetches 6 terms from Google Trends via pytrends.
Fetches one at a time with randomized delays to avoid rate limiting.
Outputs JSON to stdout; all logs go to stderr.
"""

import json
import time
import random
import sys


def fetch_term(pytrends, term: str, timeframe: str = 'today 5-y', geo: str = 'US') -> dict:
    """Fetch a single term and return {ISO_date_str: int} dict."""
    pytrends.build_payload([term], cat=0, timeframe=timeframe, geo=geo)
    df = pytrends.interest_over_time()
    if df.empty:
        print(f'[warn] Empty response for: {term}', file=sys.stderr)
        return {}
    df = df.drop(columns=['isPartial'], errors='ignore')
    return {str(k): int(v) for k, v in df[term].items()}


def main():
    try:
        from pytrends.request import TrendReq
    except ImportError:
        print('{"error": "pytrends not installed"}')
        sys.exit(1)

    pytrends = TrendReq(
        hl='en-US',
        tz=360,
        timeout=(10, 25),
        retries=3,
        backoff_factor=0.5,
    )

    terms = [
        'golf clubs',
        'golf balls',
        'golf bags',
        'golf',
        'golf equipment',
        'golf simulator',
    ]

    results = {}

    for i, term in enumerate(terms):
        print(f'[fetch] ({i+1}/{len(terms)}) {term}', file=sys.stderr)
        try:
            data = fetch_term(pytrends, term)
            results[term] = data
            print(f'[fetch] OK — {len(data)} data points', file=sys.stderr)
        except Exception as e:
            print(f'[fetch] ERROR for {term!r}: {e}', file=sys.stderr)
            results[term] = {'error': str(e)}

        # Randomized delay — never batch to avoid 429s
        if i < len(terms) - 1:
            delay = random.uniform(2.5, 4.5)
            print(f'[fetch] Sleeping {delay:.1f}s…', file=sys.stderr)
            time.sleep(delay)

    print(json.dumps(results))


if __name__ == '__main__':
    main()
